arangod/IResearch/IResearchDataStore.cpp

```diff
@@ -550,53 +589,152 @@ void ConsolidationTask::operator()() {
     TRI_ASSERT(state != nullptr);
-
-    // check state of the top-most transaction only
-    if (!state) {
-      return;  // NOOP
+    if (!state) {  // check state of the top-most transaction only
+      return;
     }
-
-    auto prev = state->cookie(this, nullptr);  // get existing cookie
-
-    if (prev) {
-      // TODO FIXME find a better way to look up a ViewState
-      auto& ctx = basics::downCast<IResearchTrxState>(*prev);
-      if (transaction::Status::COMMITTED != status) {  // rollback
-        ctx.reset();
-      } else {
-        uint64_t const lastOperationTick{state->lastOperationTick()};
-        ctx._ctx.SetLastTick(lastOperationTick);
-        if (ADB_LIKELY(!_engine->inRecovery())) {
-          ctx._ctx.SetFirstTick(lastOperationTick -
-                                state->numPrimitiveOperations());
+    auto prev = state->cookie(this);  // get existing cookie
+    if (!prev) {
+      return;
+    }
+    // TODO FIXME find a better way to look up a ViewState
+    auto& ctx = basics::downCast<IResearchTrxState>(*prev);
+#if ARANGODB_ENABLE_MAINTAINER_MODE && ARANGODB_ENABLE_FAILURE_TESTS
+    TRI_IF_FAILURE("ArangoSearch::ThreeTransactionsMisorder") {
+      size_t myNumber{0};
+      while (true) {
+        std::unique_lock sync(_t3FailureSync);
+        if (_t3PreCommit < 3) {
+          if (!myNumber) {
+            myNumber = ++_t3PreCommit;
+            LOG_TOPIC("76f6a", DEBUG, TOPIC)
+                << myNumber << " arrived to preCommit";
+          } else {
+            sync.unlock();
+            std::this_thread::sleep_for(100ms);
+          }
+        } else {
+          if (myNumber == 2) {
+            // Number 2 should go tho another flush context
+            // so wait for commit to start and
+            // then give writer some time to switch flush context
+            if (!_t3CommitSignal) {
+              sync.unlock();
+              std::this_thread::sleep_for(100ms);
+              continue;
+            }
+            std::this_thread::sleep_for(5s);
+          }
+          ctx._ctx.AddToFlush();
+          LOG_TOPIC("cdc04", DEBUG, TOPIC) << myNumber << " added to flush";
+          ++_t3NumFlushRegistered;
+          break;
         }
       }
+      // this transaction is flushed. Need to wait for all to arrive
+      while (true) {
+        std::unique_lock sync{_t3FailureSync};
+        if (_t3NumFlushRegistered == 3) {
+          // all are fluhing. Now release in order 1 2 3
+          sync.unlock();
+          std::this_thread::sleep_for(std::chrono::seconds(myNumber));
+          LOG_TOPIC("0a90a", DEBUG, TOPIC)
+              << myNumber << " released  thread " << std::this_thread::get_id();
+          break;
+        } else {
+          sync.unlock();
+          std::this_thread::sleep_for(50ms);
+        }
+      }
+    }
+    else {
+      ctx._ctx.AddToFlush();
+    }
+#else
+    ctx._ctx.AddToFlush();
+#endif
+  };
+  _trxAfterFinish = [this](transaction::Methods& trx,
```

Comment:
NIT This is called on every status change. Not only on commit.
```suggestion
  _trxStatusChange = [this](transaction::Methods& trx,
```

Reply:
In general probably name same as callback type is good, I agree